rocAL/rocAL/source/pipeline/tensor.cpp

```diff
@@ -0,0 +1,289 @@
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <cstdio>
+#if !ENABLE_HIP
+#include <CL/cl.h>
+#endif
+#include <vx_ext_amd.h>
+
+#include <cstring>
+#include <stdexcept>
+
+#include "commons.h"
+#include "tensor.h"
+
+vx_enum vx_mem_type(RocalMemType mem) {
+    switch (mem) {
+        case RocalMemType::OCL:
+            return VX_MEMORY_TYPE_OPENCL;
+        case RocalMemType::HOST:
+            return VX_MEMORY_TYPE_HOST;
+        case RocalMemType::HIP:
+            return VX_MEMORY_TYPE_HIP;
+        default:
+            throw std::runtime_error("Memory type not valid");
+    }
+}
+
+vx_size tensor_data_size(RocalTensorDataType data_type) {
+    switch (data_type) {
+        case RocalTensorDataType::FP32:
+            return sizeof(vx_float32);
+        case RocalTensorDataType::FP16:
+            return sizeof(vx_int16);
+        case RocalTensorDataType::UINT8:
+            return sizeof(vx_uint8);
+        case RocalTensorDataType::UINT32:
+            return sizeof(vx_uint32);
+        case RocalTensorDataType::INT32:
+            return sizeof(vx_int32);
+        default:
+            throw std::runtime_error("tensor data_type not valid");
+    }
+}
+
+//! Converts the Rocal data_type to OpenVX
+vx_enum interpret_tensor_data_type(RocalTensorDataType data_type) {
+    switch (data_type) {
+        case RocalTensorDataType::FP32:
+            return VX_TYPE_FLOAT32;
+        case RocalTensorDataType::FP16:
+            return VX_TYPE_FLOAT16;
+        case RocalTensorDataType::UINT8:
+            return VX_TYPE_UINT8;
+        default:
+            THROW("Unsupported Tensor type " + TOSTR(data_type))
+    }
+}
+
+bool operator==(const rocalTensorInfo &rhs, const rocalTensorInfo &lhs) {
+    return (rhs.dims() == lhs.dims() &&
+            rhs.mem_type() == lhs.mem_type() &&
+            rhs.data_type() == lhs.data_type() &&
+            rhs.color_format() == lhs.color_format() &&
+            rhs.layout() == lhs.layout());
+}
+
+void rocalTensorInfo::reallocate_tensor_roi_buffers() {
+    _roi = std::make_shared<std::vector<RocalROI>>(_batch_size);
+
+    if (_roi->size()) _roi->clear();
+    _roi->resize(_batch_size);
+    if (_is_image) {
+        for (unsigned i = 0; i < _batch_size; i++) {
+            _roi->at(i).x1 = 0;
```

Comment:
The tensor ROI will be reset each time the dimension of the tensor is changed.
In case of resize/crop the output tensor dimensions will be different from input  tensor dimensions, hence we use the API to reset the ROI of that tensor  wrt the new dimensions.